### ISSUE TITLE
Segfault in box_stacks-MoreauJeanGOSI-withfcliboutput.py

### ISSUE THREAD

**Issue by vacary** [siconos/siconos#240]
This example:
https://github.com/siconos/siconos/blob/master/examples/Mechanics/GeometricPrimitives/box_stacks-MoreauJeanGOSI-withfcliboutput.py
produces a segfault in swapInMemory. 

It seems that some interactions are not correctly initialized.

**Comment by vacary** [comment]
By the way, I opened a new PR on the fact that we update Interactions in the initialization of the Newton loop #241

**Comment by vacary** [comment]
Thanks @batzkass
The same segfault appears on other examples, for instance  
https://github.com/siconos/siconos/blob/master/examples/Mechanics/RigidBodiesCollection/n_cubes.py

**Comment by radarsat1** [comment]
Reviewing this, I see that the function `updateContactsFromChangeLog()` was added in a65e840213fc71bcad926f6733b4f2ef3e446739 among a bunch of other things.

What is the reason to separate this to another function? Apart from the fact that it's weird to mention "contacts" outside of mechanics, the change is quite wrong:

```
diff --git a/kernel/src/simulationTools/TimeStepping.cpp b/kernel/src/simulationTools/Tim
eStepping.cpp
index a933c23ba..0ddba4efc 100644
--- a/kernel/src/simulationTools/TimeStepping.cpp
+++ b/kernel/src/simulationTools/TimeStepping.cpp
@@ -373,6 +373,22 @@ void TimeStepping::computeOneStep()
   advanceToEvent();
 }
 
+
+void TimeStepping::updateContactsFromChangeLog()
+{
+  NonSmoothDynamicalSystem::ChangeLogIter p = _nsds->changeLogPosition();
+  // If graph changed, need to initialize the new interactions
+  if (p.it != _nsds->changeLogPosition().it)
+    {
+      if (initializeNSDSChangelog())
+       {
+         initOSNS();
+         _nsds->topology()->setHasChanged(true);
+       }
+    }
+}
+
+
 void TimeStepping::initializeNewtonLoop()
 {
   DEBUG_BEGIN("TimeStepping::initializeNewtonLoop()\n");
@@ -389,18 +405,8 @@ void TimeStepping::initializeNewtonLoop()
 
   // Predictive contact -- update initial contacts after updating DS positions
   {
-    NonSmoothDynamicalSystem::ChangeLogIter p = _nsds->changeLogPosition();
     updateInteractions();
-
-    // If graph changed, need to initialize the new interactions
-    if (p.it != _nsds->changeLogPosition().it)
-    {
-      if (initializeNSDSChangelog())
-      {
-        initOSNS();
-        _nsds->topology()->setHasChanged(true);
-      }
-    }
+    updateContactsFromChangeLog();
   }
 
   SP::InteractionsGraph indexSet0 = _nsds->topology()->indexSet0();
```

It can be seen that the code no longer correctly performs the check as to whether there are new entries in the changelog, instead uselessly just reading the position twice and checking if it is the same.  Then, the fix in 18fc39c13c31cd35ea19c4581f5711e24ca40d7d only removes this useless check, which should not make a difference so if it's really fixing the bug then I am confused.  The removed code is:

```
  NonSmoothDynamicalSystem::ChangeLogIter p = _nsds->changeLogPosition();
  // If graph changed, need to initialize the new interactions
  if (p.it != _nsds->changeLogPosition().it) { ...
```

How could this test *ever* be false?

(Previously it was only true if there were new entries in the changelog after `updateInteractions`, which was called between the first and second line.)

**Comment by radarsat1** [comment]
> How could this test ever be false?

*true* i meant.  Okay I reverted these changes and indeed if I put updateInteractions before the `p =` line then I get a segfault in `n_cubes.py`, but if I put it back to how it was (`p =`, then `updateInteractions`) then there is no segfault.

The change in 18fc39c simply removes this check entirely and steps through the changelog regardless, which I guess should be ok too, as it's only an optimisation, but then i guess I would argue to put the code of updateContactsFromChangelog directly into updateInteractions, or rename it initialize initializeNewInteractions or something.  I'm not sure.. something in the organization here is not very clear.  Perhaps initializeNSDSChangelog should handle calling initOSNS if necessary.
